viamrobotics/viam-python-sdk
# RSDK-2506 Sessions client

Implements client-side sessions for the python sdk

## TODO

* [x] Ensure in-progress actions are cancelled if the SDK is forcefully killed

## Manual Testing

* Copy this branch locally and `cd` into the repository.
* Copy the contents of https://gist.github.com/maximpertsov/acd5d9ddaaa835a000a310341cdfb1de into `testing.py` (or run the curl script below)

```
curl https://gist.githubusercontent.com/maximpertsov/acd5d9ddaaa835a000a310341cdfb1de/raw/0fe5f13d1cafef0eb39f7bfcb59556f7f21a53af/viam-python-sessions-testing.py > testing.py
```

* Find your favorite robot with a drive-able base and replace the `URI, SECRET, BASE_NAME` constants in script above with the relevant parameters.

* Run the following commands
```
poetry add -D aioconsole  # needed for the script above
make install && poetry run python testing.py
```

* You should see output that looks like this:
```
2023-05-25 17:21:09,768         DEBUG   viam.rpc.dial (dial.py:188)     Creating new viam-rust-utils runtime
2023-05-25 17:21:09,770         DEBUG   viam.rpc.dial (dial.py:209)     Dialing <URI> using viam-rust-utils library
2023-05-25 17:21:11,492         INFO    viam.rpc.dial (dial.py:239)     Connecting to socket: <SOCKET>
2023-05-25 17:21:11,624         DEBUG   viam.sessions_client (sessions_client.py:128)   sent heartbeat successfully
Move in a direction (WASD) or Q to exit
2023-05-25 17:21:12,038         DEBUG   viam.sessions_client (sessions_client.py:128)   sent heartbeat successfully
2023-05-25 17:21:12,450         DEBUG   viam.sessions_client (sessions_client.py:128)   sent heartbeat successfully
2023-05-25 17:21:12,865         DEBUG   viam.sessions_client (sessions_client.py:128)   sent heartbeat successfully
2023-05-25 17:21:13,279         DEBUG   viam.sessions_client (sessions_client.py:128)   sent heartbeat successfully
...
```

* Pressing `W A S or D + ENTER` should drive the base without any noticeable lag.
